src/ivc/step_folding_circuit.rs

```diff
@@ -0,0 +1,401 @@
+/// methods
+///
+/// # References
+/// - For a detailed understanding of IVC and the context in which a struct
+///   [`StepFoldingCircuit`] might be used, refer to the 'Section 5' of
+///   [Nova Whitepaper](https://eprint.iacr.org/2023/969.pdf).
+///   This trait is representation of `F'` function at 'Figure 4'
+///       - `F'` is an augmented version of `F` designed to produce the IVC proofs `P_i` at each step.
+///         It takes additional parameters such as \( vk, Ui, ui, (i, z0, zi), \omega_i, T \) and outputs \( x \).
+///
+/// - For `F'` please look at [`StepFoldingCircuit`]
+pub(crate) struct StepFoldingCircuit<'link, const ARITY: usize, C, SC, RO, const T: usize>
+where
+    C: CurveAffine,
+    C::Base: ff::PrimeFieldBits + ff::FromUniformBytes<64>,
+    C::Scalar: ff::PrimeFieldBits + ff::FromUniformBytes<64>,
+    SC: StepCircuit<ARITY, C::Base> + Sized,
+    RO: ROCircuitTrait<C::Base>,
+{
+    step_circuit: SC,
+    input: StepInputs<'link, ARITY, C, RO>,
+}
+
+impl<'link, const ARITY: usize, C, SC, RO, const T: usize> StepCircuit<ARITY, C::Base>
+    for StepFoldingCircuit<'link, ARITY, C, SC, RO, T>
+where
+    C: CurveAffine,
+    C::Base: ff::PrimeFieldBits + ff::FromUniformBytes<64>,
+    C::Scalar: ff::PrimeFieldBits + ff::FromUniformBytes<64>,
+    SC: StepCircuit<ARITY, C::Base> + Sized,
+    RO: ROCircuitTrait<C::Base, Config = MainGateConfig<T>>,
+{
+    type Config = StepConfig<ARITY, C::Base, SC, T>;
+    type FloorPlanner = SimpleFloorPlanner;
+
+    /// Configure the step circuit. This method initializes necessary
+    /// fixed columns and advice columns, but does not create any instance
+    /// columns.
+    ///
+    /// This setup is crucial for the functioning of the IVC-based system.
+    fn configure(cs: &mut ConstraintSystem<C::Base>) -> Self::Config {
+        let before = cs.num_instance_columns();
+
+        let main_gate_config = MainGate::configure(cs);
+        let step_config = <SC as StepCircuit<ARITY, C::Base>>::configure(cs);
+
+        if before != cs.num_instance_columns() {
+            panic!("You can't use instance column");
+        }
+
+        let instance = cs.instance_column();
+        cs.enable_equality(instance);
+        StepConfig {
+            instance,
+            step_config,
+            main_gate_config,
+        }
+    }
+
+    /// Sythesize the circuit for a computation step and return variable
+    /// that corresponds to the output of the step z_{i+1}
+    /// this method will be called when we synthesize the IVC_Circuit
+    ///
+    /// Return `z_out` result
+    fn synthesize_step(
+        &self,
+        config: Self::Config,
+        layouter: &mut impl Layouter<C::Base>,
+        assigned_z_i: &[AssignedCell<C::Base, C::Base>; ARITY],
+    ) -> Result<[AssignedCell<C::Base, C::Base>; ARITY], SynthesisError> {
+        let assigned_z_0: [_; ARITY] = layouter.assign_region(
```

Comment:
Instead of assign `z_0`, we can use `assign_step_inputs` to assign all fields in `StepInputs`.

Reply:
It's more complicated & because of that I separate issue for this https://github.com/snarkify/sirius/issues/153